arch/arc/core/cache.c

```diff
@@ -111,35 +96,35 @@ static void dcache_flush_mlines(uint32_t start_addr, uint32_t size)
 			}
 		} while (1);
 		start_addr += DCACHE_LINE_SIZE;
-	} while (start_addr <= end_addr);
+	} while (start_addr < end_addr);
 
 	arch_irq_unlock(key); /* --exit critical section-- */
 
 }
 
+void arch_dcache_invd(void *start_addr_ptr, size_t size)
+{
+	uintptr_t start_addr = (uintptr_t)start_addr_ptr;
+	uintptr_t end_addr;
+	unsigned int key;
 
-/**
- *
- * @brief Flush d-cache lines to main memory
- *
- * No alignment is required for either <virt> or <size>, but since
- * sys_cache_flush() iterates on the d-cache lines, a d-cache line alignment for
- * both is optimal.
- *
- * The d-cache line size is specified either via the CONFIG_CACHE_LINE_SIZE
- * kconfig option or it is detected at runtime.
- *
- * @param start_addr the pointer to start the multi-line flush
- * @param size the number of bytes that are to be flushed
- *
- * @return N/A
- */
+	if (!dcache_available() || (size == 0U)) {
+		return;
+	}
+	end_addr = start_addr + size;
+	start_addr = ROUND_DOWN(start_addr, DCACHE_LINE_SIZE);
 
-void sys_cache_flush(vaddr_t start_addr, size_t size)
-{
-	dcache_flush_mlines((uint32_t)start_addr, (uint32_t)size);
-}
+	key = irq_lock(); /* -enter critical section- */
```

Comment:
The flush routine uses arch_irq_lock (in case of SMP: local CPU interrupt disable, not a global spinlock). That would be enough here as well.
(and, there's a tradeoff between locking overhead and interrupt blocking time: the lock is only required within the loop. If small regions are flushed, this is better - let's leave it like this)

Reply:
to avoid confusion: can you change the irq_lock to arch_irq_lock (the "let's leave it like this" referred to moving the lock into the loop).

Reply:
Changed it to arch_irq_lock to be consistent as well , Thank you!!